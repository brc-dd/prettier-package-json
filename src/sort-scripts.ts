import sortObject from 'sort-object-keys';
import orderBy from 'sort-order';
import { PackageJson } from './types';

const checkPre = (arg: string, scripts: PackageJson['scripts']): boolean =>
  /^pre/.test(arg) && scripts!.hasOwnProperty(arg.substr(3));

const checkPost = (arg: string, scripts: PackageJson['scripts']): boolean =>
  /^post/.test(arg) && scripts!.hasOwnProperty(arg.substr(4));

// Sort alphabetically by script name excluding pre/post prefixes
function scriptName(this: PackageJson['scripts'], ...args: [string, string]): 1 | 0 | -1 {
  const [a, b] = args.map((arg) =>
    checkPre(arg, this) || checkPost(arg, this) ? arg.replace(/^(pre|post)/, '') : arg,
  );
  return a === b ? 0 : a < b ? -1 : 1;
}

// Sort by pre, script, post
function prePostHooks(this: PackageJson['scripts'], a: string, b: string): 1 | 0 | -1 {
  if (checkPre(a, this) || checkPost(b, this)) return -1;
  else if (checkPost(a, this) || checkPre(b, this)) return 1;
  return 0;
}

export default function sortScripts(scripts: PackageJson['scripts'] = {}): {
  scripts?: PackageJson['scripts'];
} {
  const order = orderBy(scriptName.bind(scripts), prePostHooks.bind(scripts));
  const keys = Object.keys(scripts) as never[];
  return keys.length === 0 ? {} : { scripts: sortObject(scripts, keys.sort(order)) };
}
