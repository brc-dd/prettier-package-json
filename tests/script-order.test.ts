import { format } from '../src';

test('It orders scripts in alphabetical order, keeping pre and post scripts beside their counterparts', () => {
  const json = {
    scripts: {
      test: 'test',
      pretest: 'pretest',
      version: 'version',
      postversion: 'postversion',
      build: 'build',
    },
  };

  expect(format(json)).toMatchSnapshot();
});

test('It sorts scripts beginning with pre or post in alphabetical order if their base script is not present', () => {
  const json = {
    scripts: {
      prettify: 'prettify',
      prepare: 'prepare',
      parse: 'parse',
      postprettify: 'postprettify',
      prestart: 'prestart',
      preprettify: 'preprettify',
      start: 'start',
      'postcss:watch': 'postcss:watch',
    },
  };

  expect(format(json)).toMatchSnapshot();
});
