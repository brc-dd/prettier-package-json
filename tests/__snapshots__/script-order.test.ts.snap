// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It orders scripts in alphabetical order, keeping pre and post scripts beside their counterparts 1`] = `
"{
  \\"scripts\\": {
    \\"build\\": \\"build\\",
    \\"pretest\\": \\"pretest\\",
    \\"test\\": \\"test\\",
    \\"version\\": \\"version\\",
    \\"postversion\\": \\"postversion\\"
  }
}
"
`;

exports[`It sorts scripts beginning with pre or post in alphabetical order if their base script is not present 1`] = `
"{
  \\"scripts\\": {
    \\"parse\\": \\"parse\\",
    \\"postcss:watch\\": \\"postcss:watch\\",
    \\"prepare\\": \\"prepare\\",
    \\"preprettify\\": \\"preprettify\\",
    \\"prettify\\": \\"prettify\\",
    \\"postprettify\\": \\"postprettify\\",
    \\"prestart\\": \\"prestart\\",
    \\"start\\": \\"start\\"
  }
}
"
`;
